src/mapreduce.jl

```diff
@@ -285,3 +285,62 @@ end
         @inbounds return similar_type(a, T, Size($Snew))(tuple($(exprs...)))
     end
 end
+
+struct _InitialValue end
+
+_maybe_val(dims::Integer) = Val(Int(dims))
+_maybe_val(dims) = dims
+_valof(::Val{D}) where D = D
+
+@inline Base.accumulate(op::F, a::StaticVector; dims = :, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybe_val(dims), init)
+
+@inline Base.accumulate(op::F, a::StaticArray; dims, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybe_val(dims), init)
+
+@inline function _accumulate(op::F, a::StaticArray, dims::Union{Val,Colon}, init) where {F}
+    if isempty(a)
+        if init isa _InitialValue
+            # Deliberately not using `return_type` here, since this `eltype` is
+            # exact for the singleton element case (i.e., `op` will not be called).
+            return similar_type(a)()
+        else
+            # Using the type that _would_ be used if `size(a, dims) == 1`:
+            T = return_type(op, Tuple{typeof(init), eltype(a)})
+            return similar_type(a, T)()
+        end
+    end
+
+    # Adjoin the initial value to `op`:
+    rf(x, y) = x isa _InitialValue ? y : op(x, y)
+
+    # StaticArrays' `reduce` is `foldl`:
+    results = _reduce(
+        a,
+        dims,
+        (init = (similar_type(a, Union{}, Size(0))(), init),),
+    ) do (ys, acc), x
+        y = rf(acc, x)
+        # Not using `push(ys, y)` here since we need to widen element type as
+        # we iterate.
+        (vcat(ys, SA[y]), y)
+    end
+    dims === (:) && return first(results)
+
+    ys = map(first, results)
+    # Now map over all indices of `a`.  Since `_map` needs at least
+    # one `StaticArray` to be passed, we pass `a` here, even though
+    # the values of `a` are not used.
+    data = _map(a, CartesianIndices(a)) do _, CI
+        D = _valof(dims)
+        I = Tuple(CI)
+        J = setindex(I, 1, D)
+        ys[J...][I[D]]
+    end
+    return similar_type(a, eltype(data))(data)
+end
+
+@inline Base.cumsum(a::StaticArray; kw...) =
+    accumulate(Base.add_sum, a; init = Base.reduce_empty(Base.add_sum, eltype(a)), kw...)
+@inline Base.cumprod(a::StaticArray; kw...) =
+    accumulate(Base.mul_prod, a; init = Base.reduce_empty(Base.mul_prod, eltype(a)), kw...)
```

Comment:
Re: `cumsum(SA{Int8}[1])` https://github.com/JuliaArrays/StaticArrays.jl/pull/702#discussion_r379965085

It's easy to make it `Base`-compatible. So maybe it's worth doing this?

Reply:
Oh I guess it's not so bad for `+` given that already has a specialized `add_sum`. So sure, if it's easy let's make this compatible.